README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

[![Project Status: Active – The project has reached a stable, usable
state and is being actively
developed.](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active)
[![Travis build
status](https://travis-ci.org/HealthEconomicsHackathon/inflatably.svg?branch=use_travis)](https://travis-ci.org/HealthEconomicsHackathon/inflatably)
[![codecov.io](https://codecov.io/github/HealthEconomicsHackathon/inflatably/coverage.svg?branch=master)](https://codecov.io/github/HealthEconomicsHackathon/inflatably?branch=master)

# inflatably

![Alt Text](giphy.gif)

inflatably is an R package [available on
GitHub](https://github.com/HealthEconomicsHackathon/inflatably).

The inflatably can also be installed from GitHub:

``` r
# install.packages("devtools")
devtools::install_github("HealthEconomicsHackathon/inflatably")
```

-----

# Motivation

Inflating costs to present value (PV) is a routine step in
cost-effectiveness analyses. Further, there are different inflation
rates depending on the type of cost to be inflated.

## License

MIT © Imperial College of Science, Technology and Medicine


NEWS.md
# inflatably 0.1.0

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(inflatably)

test_check("inflatably")


tests/testthat/test-inflation_adjust_cost_data.R
context("test-inflation_adjust_cost_data")

test_that("simple calcs", {
  expect_equivalent(inflation_adjust_cost_data(2014, 2015, 1, "HCHS"), 2.7)
  # expect_equivalent(inflation_adjust_cost_data(2014, 2015, 1, "CPI"), )
  
})

test_that("errors", {
})

inflation_df <- data.frame("year" = 2008:2010,
                           "rate" = c(0.02, 0.03, 0.025))

test_that("asserts custom", {
  expect_error(inflation_adjust_cost_custom(2008.5, 2010, 1, inflation_df),
               regexp = "From date must be an integer valued whole year")
  
  expect_error(inflation_adjust_cost_custom(2008, 2009.5, 1, inflation_df),
               regexp = "To date must be an integer valued whole year")
  
  expect_error(inflation_adjust_cost_custom(2008, 2010, -1, inflation_df),
               regexp = "Cost must be non-negative")
  
  expect_error(inflation_adjust_cost_custom(1008, 2010, 1, inflation_df),
               regexp = "from year not in look-up table")
  
  expect_error(inflation_adjust_cost_custom(2008, 2020, 1, inflation_df),
               regexp = "to year not in look-up table")
})

test_that("asserts main", {
  
  expect_error(inflation_adjust_cost_data(2012, 2015, 2, "abc"),
               "inflation_data name not available")
})

R/data-inflation.R

#' PSSRU annual inflation hospital and community health services (HCHS)
#'
#' Until 2016/2017, a hospital & community health services (HCHS) index was calculated
#' by the Department of Health (now DHSC). The hospital and community health services (HCHS)
#' pay and price inflation was a weighted average of two separate inflation indices: the pay
#' cost index was calculated using the annual increase in NHS salaries and the Health Service
#' Cost Index (HSCI) measured the price change for each of 40 sub-indices of goods and
#' services purchased by the HCHS. These were weighted according to the proportion of
#' expenditure on pay and prices to give the HCHS pay and prices index.
#' In 2016, this index was discontinued, and in 2018, in collaboration with the DHSC,
#' PSSRU have developed a new inflation index.
#'
#' @format data.frame with a row for each year. From-to July. Variables are:
#' \describe{
#'    \item{year}{to year}
#'    \item{rate}{Proportion change from previous year}
#' }
#'
#' @source 2017 \url{https://www.pssru.ac.uk/pub/uc/uc2017/sources-of-information.pdf}
"HCHS"


#' Consumer Price Inflation (CPI) Medicinal Products
#'
#' @format data.frame with a row for each year. Variables are:
#' \describe{
#'    \item{year}{to year}
#'    \item{rate}{Proportion change from previous year}
#' }
#'
#' @source \url{https://www.ons.gov.uk/economy/inflationandpriceindices/datasets/consumerpriceinflation}
"CPI"

R/inflation_adjust_costs.R


#' Calculate annual inflation adjusted costs
#'
#' Up to a present time inflated upwards.
#'
#' Option to use the following datasets:
#'
#' \itemize{
#'   \item \href{https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/562750/GDP_Deflators_Qtrly_National_Accounts_September_2016_update_v2.csv}{ONS GDP_Deflators_Qtrly_National_Accounts}
#' (This document contains the latest gross domestic product (GDP) deflators.
#' The GDP deflator can be viewed as a measure of general inflation in the domestic economy)
#'   \item \href{https://www.pssru.ac.uk/pub/uc/uc2017/sources-of-information.pdf}{PSSRU annual inflation hospital and community health services (HCHS)}
#'   \item \href{https://www.ons.gov.uk/economy/inflationandpriceindices/datasets/consumerpriceinflation}{Consumer Price Inflation (CPI)}
#'   \item Fixed 3.5\%.
#' }
#'
#' \deqn{(1 + i_{1})(1 + i_{2}) \cdots (1 + i_{n}) \times C}
#'
#' Clearly, for the same \code{i} each year this is simply
#' \deqn{C(1 + i)^n}
#'
#' @param from_year Date of cost to convert from
#' @param to_year Date to convert cost to
#' @param from_cost Cost at \code{from_year}
#' @param inflation_data NA default is fixed 3.5\% rate of inflation; otherwise source of data (string) \code{\link{GDP_deflators}}, \code{\link{HCHS_pay}}, \code{\link{{HCHS_price}} or \code{\link{CPI}}.
#'
#' @return Inflated (to) cost (scalar), with attributes used to generate the return value:
#' \itemize{
#'   \item \code{from_year}
#'   \item \code{to_year}
#'   \item \code{from_cost}
#'   \item \code{reference} Which inflation adjustment dataset used.
#'  }
#' @export
#'
#' @examples
#' from_year <- 2012
#' to_year <- 2015
#' from_cost <- 96.140
#'
# inflation_adjust_cost(from_year,
#                       to_year,
#                       from_cost)
#' #100
#'
#' inflation_adjust_cost(from_year = 2010,
#'                       to_year = 2016,
#'                       from_cost = 1)
#' #1.229255
#' 1*(1+0.035)^6
#'
#' inflation_adjust_cost(from_year = 2014,
#'                       to_year = 2016,
#'                       from_cost = 1,
#'                       inflation_data = "CPI")
#' 1*1.004*1.01
#'
inflation_adjust_cost <- function(from_year,
                                  to_year,
                                  from_cost,
                                  inflation_data = 0.035){
  
  ##TODO##
  # the webpage deflator file is too messy- with comments etc- to use as raw data
  # # how to always use the latest?
  # coulud use readxl but file layout may change
  # temp <- tempfile()
  # download.file("https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/562750/GDP_Deflators_Qtrly_National_Accounts_September_2016_update_v2.csv", temp)
  # https://www.gov.uk/government/statistics/gdp-deflators-at-market-prices-and-money-gdp-march-2018-quarterly-national-accounts
  # deflators <- read_csv(temp)
  # unlink(temp)
  
  if (from_year %% 1 != 0) stop("From date must be an integer valued whole year")
  if (to_year %% 1 != 0) stop("To date must be an integer valued whole year")
  if (from_cost < 0) stop("Cost must be non-negative")
  
  data_sources <- dir(system.file("extdata", package = "QALY"))
  
  if (!(inflation_data %in% data_sources || is_prob(inflation_data))) {
    
    stop("inflation_data not allowed")
  }
  
  num_years <- to_year - from_year
  
  if (inflation_data %in% data_sources) {
    
    data_files <- system.file("extdata", inflation_data, package = "QALY")
    datapkg_data <- datapkg::datapkg_read(data_files)
    data <- datapkg_data$data[[inflation_data]]
    data$adjustment <- data$prop_change_from_prev_year
    from_row <- which(data$year == from_year)
    
  } else {
    
    data <- NULL
    data$adjustment <- rep(inflation_data, num_years)
    from_row <- 0
  }
  
  to_cost <- from_cost
  
  for (i in seq_len(num_years)) {
    
    to_cost <- to_cost * (1 + data$adjustment[from_row + i])
  }
  
  attr(to_cost, "from_year") <- from_year
  attr(to_cost, "to_year")   <- to_year
  attr(to_cost, "from_cost") <- from_cost
  attr(to_cost, "reference") <- inflation_data
  
  return(to_cost)
}

#'@param from_year base year to inflate from
#'@param to_year destination year to inflate to
#'@param from_cost base year cost 
#'@param inflation_df dataframe of annual inflation data, in the format
#'"year" = year (yyyy) ? "rate" = annual inflation rate (1/100th fraction)
#'
#'@examples 
#'inflation_df <- data.frame("year" = 2008:2010, "rate"=c(0.02, 0.03, 0.025))
#'inflation_adjust_cost_custom(2008,2009,100,inflation_df)
#'Result: 100*1.02 = 102 
inflation_adjust_cost_custom <- function(from_year,
                                         to_year,
                                         from_cost,
                                         inflation_df) {
  
  return_cost <- from_cost
  
  for (i in from_year:(to_year-1)) {
    rate <- inflation_df[which(inflation_df$year==2008),"rate"]
    return_cost <- return_cost * (1+rate)
  }
  
 return_cost
}


R/inflation_adjust_costs_data.R


#' Calculate annual inflation adjusted costs
#'
#' Up to a present value (PV) inflated upwards.
#'
#' Option to use the following datasets:
#'
#' \itemize{
#'   \item \href{https://www.pssru.ac.uk/pub/uc/uc2017/sources-of-information.pdf}{PSSRU annual inflation hospital and community health services (HCHS)}
#'   \item \link{CPI}
#' }
#'
#' \deqn{(1 + i_{1})(1 + i_{2}) \cdots (1 + i_{n}) \times C}
#'
#' Clearly, for the same \code{i} each year this is simply
#' \deqn{C(1 + i)^n}
#'
#' @param from_year Date of cost to convert from
#' @param to_year Date to convert cost to
#' @param from_cost Cost at \code{from_year}
#' @param inflation_df_nm Source of data (string) \code{\link{HCHS}}, or \code{\link{CPI}}.
#'
#' @return Inflated (to) cost (scalar), with attributes used to generate the return value:
#' \itemize{
#'   \item \code{from_year}
#'   \item \code{to_year}
#'   \item \code{from_cost}
#'   \item \code{reference} Which inflation adjustment dataset used.
#'  }
#' @export
#'
#' @examples
#' from_year <- 2012
#' to_year <- 2015
#' from_cost <- 96.140
#'
# inflation_adjust_cost_data(from_year,
#                            to_year,
#                            from_cost)
#' 1*(1+0.035)^6
#'
#' inflation_adjust_cost_data(from_year = 2014,
#'                            to_year = 2016,
#'                            from_cost = 1,
#'                            inflation_df_nm = "HCHS")
#' 1*1.004*1.01
#'
inflation_adjust_cost_data <- function(from_year,
                                       to_year,
                                       from_cost,
                                       inflation_df_nm){
  
  data_sources <- utils::data(package = "inflately")
  data_sources <- data_sources$results[ ,"Item"]
  
  if (!(inflation_df_nm %in% data_sources)) {
    
    stop("inflation_data name not available")
  }

  # load data to workspace
  eval(parse(text = paste0("utils::data(", inflation_df_nm, ")")))
  inflation_df <- eval(parse(text = inflation_df_nm))
  
  to_cost <- 
    inflation_adjust_cost_custom(from_year,
                                 to_year,
                                 from_cost,
                                 inflation_df)   
  
  attr(to_cost, "from_year") <- from_year
  attr(to_cost, "to_year")   <- to_year
  attr(to_cost, "from_cost") <- from_cost
  attr(to_cost, "reference") <- inflation_df_nm
  
  return(to_cost)
}


R/inflation_adjust_cost_custom.R

#' Inflation adjust a cost to present value
#'
#' @param from_year base year to inflate from
#' @param to_year destination year to inflate to
#' @param from_cost base year cost 
#' @param inflation_df inflation annual rate data. dataframe of annual inflation data, in the format
#' "year" = year (yyyy), "rate" = annual inflation rate (1/100th fraction)
#'
#' @examples 
#' inflation_df <- data.frame("year" = 2008:2010, "rate" = c(0.02, 0.03, 0.025))
#' inflation_adjust_cost_custom(2008, 2009, 100, inflation_df)
#' 100*1.02 # 102 
#'
inflation_adjust_cost_custom <- function(from_year,
                                         to_year,
                                         from_cost,
                                         inflation_df) {
  
  stopifnot(from_year %in% inflation_df$year)
  stopifnot(to_year %in% inflation_df$year)
  if (from_year %% 1 != 0) stop("From date must be an integer valued whole year")
  if (to_year %% 1 != 0) stop("To date must be an integer valued whole year")
  if (from_cost < 0) stop("Cost must be non-negative")
  
  to_cost <- from_cost
  
  for (i in from_year:(to_year - 1)) {
    
    rate <- inflation_df[inflation_df$year == i, "rate"]
    to_cost <- to_cost * (1 + rate)
  }
  
 return(to_cost)
}


R/inflation_adjust_cost_datapkg.R


#' Calculate annual inflation adjusted costs
#'
#' Up to a present time inflated upwards.
#'
#' Option to use the following datasets:
#'
#' \itemize{
#'   \item \href{https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/562750/GDP_Deflators_Qtrly_National_Accounts_September_2016_update_v2.csv}{ONS GDP_Deflators_Qtrly_National_Accounts}
#' (This document contains the latest gross domestic product (GDP) deflators.
#' The GDP deflator can be viewed as a measure of general inflation in the domestic economy)
#'   \item \href{https://www.pssru.ac.uk/pub/uc/uc2017/sources-of-information.pdf}{PSSRU annual inflation hospital and community health services (HCHS)}
#'   \item \href{https://www.ons.gov.uk/economy/inflationandpriceindices/datasets/consumerpriceinflation}{Consumer Price Inflation (CPI)}
#'   \item Fixed 3.5\%.
#' }
#'
#' \deqn{(1 + i_{1})(1 + i_{2}) \cdots (1 + i_{n}) \times C}
#'
#' Clearly, for the same \code{i} each year this is simply
#' \deqn{C(1 + i)^n}
#'
#' @param from_year Date of cost to convert from
#' @param to_year Date to convert cost to
#' @param from_cost Cost at \code{from_year}
#' @param inflation_data NA default is fixed 3.5\% rate of inflation; otherwise source of data (string) \code{\link{GDP_deflators}}, \code{\link{HCHS_pay}}, \code{\link{{HCHS_price}} or \code{\link{CPI}}.
#'
#' @return Inflated (to) cost (scalar), with attributes used to generate the return value:
#' \itemize{
#'   \item \code{from_year}
#'   \item \code{to_year}
#'   \item \code{from_cost}
#'   \item \code{reference} Which inflation adjustment dataset used.
#'  }
#' @export
#'
#' @examples
#' from_year <- 2012
#' to_year <- 2015
#' from_cost <- 96.140
#'
# inflation_adjust_cost(from_year,
#                       to_year,
#                       from_cost)
#' #100
#'
#' inflation_adjust_cost(from_year = 2010,
#'                       to_year = 2016,
#'                       from_cost = 1)
#' #1.229255
#' 1*(1+0.035)^6
#'
#' inflation_adjust_cost(from_year = 2014,
#'                       to_year = 2016,
#'                       from_cost = 1,
#'                       inflation_data = "CPI")
#' 1*1.004*1.01
#'
inflation_adjust_cost_datapkg <- function(from_year,
                                  to_year,
                                  from_cost,
                                  inflation_data = 0.035){
  
  ##TODO##
  # the webpage deflator file is too messy- with comments etc- to use as raw data
  # # how to always use the latest?
  # coulud use readxl but file layout may change
  # temp <- tempfile()
  # download.file("https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/562750/GDP_Deflators_Qtrly_National_Accounts_September_2016_update_v2.csv", temp)
  # https://www.gov.uk/government/statistics/gdp-deflators-at-market-prices-and-money-gdp-march-2018-quarterly-national-accounts
  # deflators <- read_csv(temp)
  # unlink(temp)
  
  if (from_year %% 1 != 0) stop("From date must be an integer valued whole year")
  if (to_year %% 1 != 0) stop("To date must be an integer valued whole year")
  if (from_cost < 0) stop("Cost must be non-negative")
  
  data_sources <- dir(system.file("extdata", package = "QALY"))
  
  if (!(inflation_data %in% data_sources || is_prob(inflation_data))) {
    
    stop("inflation_data not allowed")
  }
  
  num_years <- to_year - from_year
  
  if (inflation_data %in% data_sources) {
    
    data_files <- system.file("extdata", inflation_data, package = "QALY")
    datapkg_data <- datapkg::datapkg_read(data_files)
    data <- datapkg_data$data[[inflation_data]]
    data$adjustment <- data$prop_change_from_prev_year
    from_row <- which(data$year == from_year)
    
  } else {
    
    data <- NULL
    data$adjustment <- rep(inflation_data, num_years)
    from_row <- 0
  }
  
  to_cost <- from_cost
  
  for (i in seq_len(num_years)) {
    
    to_cost <- to_cost * (1 + data$adjustment[from_row + i])
  }
  
  attr(to_cost, "from_year") <- from_year
  attr(to_cost, "to_year")   <- to_year
  attr(to_cost, "from_cost") <- from_cost
  attr(to_cost, "reference") <- inflation_data
  
  return(to_cost)
}

R/inflation_adjust_costs_datapkg.R


#' Calculate annual inflation adjusted costs
#'
#' Up to a present time inflated upwards.
#'
#' Option to use the following datasets:
#'
#' \itemize{
#'   \item \href{https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/562750/GDP_Deflators_Qtrly_National_Accounts_September_2016_update_v2.csv}{ONS GDP_Deflators_Qtrly_National_Accounts}
#' (This document contains the latest gross domestic product (GDP) deflators.
#' The GDP deflator can be viewed as a measure of general inflation in the domestic economy)
#'   \item \href{https://www.pssru.ac.uk/pub/uc/uc2017/sources-of-information.pdf}{PSSRU annual inflation hospital and community health services (HCHS)}
#'   \item \href{https://www.ons.gov.uk/economy/inflationandpriceindices/datasets/consumerpriceinflation}{Consumer Price Inflation (CPI)}
#'   \item Fixed 3.5\%.
#' }
#'
#' \deqn{(1 + i_{1})(1 + i_{2}) \cdots (1 + i_{n}) \times C}
#'
#' Clearly, for the same \code{i} each year this is simply
#' \deqn{C(1 + i)^n}
#'
#' @param from_year Date of cost to convert from
#' @param to_year Date to convert cost to
#' @param from_cost Cost at \code{from_year}
#' @param inflation_data NA default is fixed 3.5\% rate of inflation; otherwise source of data (string) \code{\link{GDP_deflators}}, \code{\link{HCHS_pay}}, \code{\link{{HCHS_price}} or \code{\link{CPI}}.
#'
#' @return Inflated (to) cost (scalar), with attributes used to generate the return value:
#' \itemize{
#'   \item \code{from_year}
#'   \item \code{to_year}
#'   \item \code{from_cost}
#'   \item \code{reference} Which inflation adjustment dataset used.
#'  }
#' @export
#'
#' @examples
#' from_year <- 2012
#' to_year <- 2015
#' from_cost <- 96.140
#'
# inflation_adjust_cost(from_year,
#                       to_year,
#                       from_cost)
#' #100
#'
#' inflation_adjust_cost(from_year = 2010,
#'                       to_year = 2016,
#'                       from_cost = 1)
#' #1.229255
#' 1*(1+0.035)^6
#'
#' inflation_adjust_cost(from_year = 2014,
#'                       to_year = 2016,
#'                       from_cost = 1,
#'                       inflation_data = "CPI")
#' 1*1.004*1.01
#'
inflation_adjust_cost_datapkg <- function(from_year,
                                  to_year,
                                  from_cost,
                                  inflation_data = 0.035){
  
  ##TODO##
  # the webpage deflator file is too messy- with comments etc- to use as raw data
  # # how to always use the latest?
  # coulud use readxl but file layout may change
  # temp <- tempfile()
  # download.file("https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/562750/GDP_Deflators_Qtrly_National_Accounts_September_2016_update_v2.csv", temp)
  # https://www.gov.uk/government/statistics/gdp-deflators-at-market-prices-and-money-gdp-march-2018-quarterly-national-accounts
  # deflators <- read_csv(temp)
  # unlink(temp)
  
  if (from_year %% 1 != 0) stop("From date must be an integer valued whole year")
  if (to_year %% 1 != 0) stop("To date must be an integer valued whole year")
  if (from_cost < 0) stop("Cost must be non-negative")
  
  data_sources <- dir(system.file("extdata", package = "QALY"))
  
  if (!(inflation_data %in% data_sources || is_prob(inflation_data))) {
    
    stop("inflation_data not allowed")
  }
  
  num_years <- to_year - from_year
  
  if (inflation_data %in% data_sources) {
    
    data_files <- system.file("extdata", inflation_data, package = "QALY")
    datapkg_data <- datapkg::datapkg_read(data_files)
    data <- datapkg_data$data[[inflation_data]]
    data$adjustment <- data$prop_change_from_prev_year
    from_row <- which(data$year == from_year)
    
  } else {
    
    data <- NULL
    data$adjustment <- rep(inflation_data, num_years)
    from_row <- 0
  }
  
  to_cost <- from_cost
  
  for (i in seq_len(num_years)) {
    
    to_cost <- to_cost * (1 + data$adjustment[from_row + i])
  }
  
  attr(to_cost, "from_year") <- from_year
  attr(to_cost, "to_year")   <- to_year
  attr(to_cost, "from_cost") <- from_cost
  attr(to_cost, "reference") <- inflation_data
  
  return(to_cost)
}



R/inflation_adjust_costs_custom.R

#' Inflation adjust a cost to present value
#'
#' @param from_year base year to inflate from
#' @param to_year destination year to inflate to
#' @param from_cost base year cost 
#' @param inflation_df inflation annual rate data. dataframe of annual inflation data, in the format
#' "year" = year (yyyy), "rate" = annual inflation rate (1/100th fraction)
#'
#' @export
#' 
#' @examples 
#' inflation_df <- data.frame("year" = 2008:2010, "rate" = c(0.02, 0.03, 0.025))
#' inflation_adjust_cost_custom(2008, 2009, 100, inflation_df)
#' 100*1.02 # 102 
#'
inflation_adjust_cost_custom <- function(from_year,
                                         to_year,
                                         from_cost,
                                         inflation_df) {
  
  if (from_year %% 1 != 0) stop("From date must be an integer valued whole year")
  if (to_year %% 1 != 0) stop("To date must be an integer valued whole year")
  if (from_cost < 0) stop("Cost must be non-negative")
  if(!from_year %in% inflation_df$year) stop("from year not in look-up table")
  if(!to_year %in% inflation_df$year) stop("to year not in look-up table")
  
  to_cost <- from_cost
  
  for (i in from_year:(to_year - 1)) {
    
    rate <- inflation_df[inflation_df$year == i, "rate"]
    to_cost <- to_cost * (1 + rate)
  }
  
  return(to_cost)
}